jetty-client/src/main/java/org/eclipse/jetty/client/Socks5Proxy.java

```diff
@@ -0,0 +1,416 @@
+     * @return the previous authentication method of the same type, or {@code null}
+     * if there was none of that type already present
+     */
+    public Socks5.Authentication.Factory putAuthenticationFactory(Socks5.Authentication.Factory authenticationFactory)
+    {
+        return authentications.put(authenticationFactory.getMethod(), authenticationFactory);
+    }
+
+    /**
+     * <p>Removes the authentication of the given {@code method}.</p>
+     *
+     * @param method the authentication method to remove
+     */
+    public Socks5.Authentication.Factory removeAuthenticationFactory(byte method)
+    {
+        return authentications.remove(method);
+    }
+
+    @Override
+    public ClientConnectionFactory newClientConnectionFactory(ClientConnectionFactory connectionFactory) 
+    {
+        return new Socks5ProxyClientConnectionFactory(connectionFactory);
+    }
+
+    private class Socks5ProxyClientConnectionFactory implements ClientConnectionFactory
+    {
+        private final ClientConnectionFactory connectionFactory;
+
+        private Socks5ProxyClientConnectionFactory(ClientConnectionFactory connectionFactory)
+        {
+            this.connectionFactory = connectionFactory;
+        }
+
+        public org.eclipse.jetty.io.Connection newConnection(EndPoint endPoint, Map<String, Object> context)
+        {
+            HttpDestination destination = (HttpDestination)context.get(HttpClientTransport.HTTP_DESTINATION_CONTEXT_KEY);
+            Executor executor = destination.getHttpClient().getExecutor();
+            Socks5ProxyConnection connection = new Socks5ProxyConnection(endPoint, executor, connectionFactory, context, authentications);
+            return customize(connection, context);
+        }
+    }
+
+    private static class Socks5ProxyConnection extends AbstractConnection implements org.eclipse.jetty.io.Connection.UpgradeFrom
+    {
+        private static final Pattern IPv4_PATTERN = Pattern.compile("(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})\\.(\\d{1,3})");
+
+        // SOCKS5 response max length is 262 bytes.
+        private final ByteBuffer byteBuffer = BufferUtil.allocate(512);
+        private final ClientConnectionFactory connectionFactory;
+        private final Map<String, Object> context;
+        private final Map<Byte, Socks5.Authentication.Factory> authentications;
+        private State state = State.HANDSHAKE;
+
+        private Socks5ProxyConnection(EndPoint endPoint, Executor executor, ClientConnectionFactory connectionFactory, Map<String, Object> context, Map<Byte, Socks5.Authentication.Factory> authentications)
+        {
+            super(endPoint, executor);
+            this.connectionFactory = connectionFactory;
+            this.context = context;
+            this.authentications = Map.copyOf(authentications);
+        }
+
+        @Override
+        public ByteBuffer onUpgradeFrom()
+        {
+            return BufferUtil.copy(byteBuffer);
```

Comment:
Before onUpgradeFrom returns, byteBuffer can be set to null so that it can be collected by GC to save memory. This is useful if user has created many connections.

Reply:
The whole connection will be GCed because it is being replaced by another.